src/shims/unix/fs.rs

```diff
@@ -652,26 +653,26 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
         // Reject if isolation is enabled.
         if let IsolatedOp::Reject(reject_with) = this.machine.isolated_op {
             this.reject_in_isolation("`stat`", reject_with)?;
-            let eacc = this.eval_libc("EACCES");
-            this.set_last_error(eacc)?;
-            return Ok(Scalar::from_i32(-1));
+            return this.set_libc_err_and_return_neg1("EACCES", dest);
         }
 
         // `stat` always follows symlinks.
-        let metadata = match FileMetadata::from_path(this, &path, true)? {
+        let metadata = match FileMetadata::from_path(this, &path, true, dest)? {
             Some(metadata) => metadata,
-            None => return Ok(Scalar::from_i32(-1)), // `FileMetadata` has set errno
+            None => return Ok(EmulateItemResult::NeedsReturn),
```

Comment:
```suggestion
            None => return Ok(EmulateItemResult::NeedsReturn), // `FileMetadata::from_path` has set the return value
```
But this is a bit odd, seems cleaner to instead have `from_path` return a `Result` and do the errno+return handling here?

Reply:
A bit annoying as I have to return a `Result<T, Scalar>` to be able to handle both EBADF and `io::Error`, but generally nicer indeed